diff --git a/sysdeps/zigux/src/sysdeps.cpp b/sysdeps/zigux/src/sysdeps.cpp
index b44fa24..f296181 100644
--- a/sysdeps/zigux/src/sysdeps.cpp
+++ b/sysdeps/zigux/src/sysdeps.cpp
@@ -7,7 +7,7 @@
 #include <mlibc/fsfd_target.hpp>
 #include <zigux/syscall.h>
 
-#define STUB_IMPL { sys_libc_log("stub function was called"); sys_libc_panic(); }
+#define STUB_IMPL(name) { sys_libc_log("stub function " #name " was called"); sys_libc_panic(); }
 
 namespace mlibc {
 
@@ -24,30 +24,18 @@ void sys_libc_log(const char *message) {
     sys_exit(1);
 }
 
-int sys_tcb_set(void *pointer) STUB_IMPL
-
-int sys_futex_wait(int *pointer, int expected, const struct timespec *time) STUB_IMPL
-
-int sys_futex_wake(int *pointer) STUB_IMPL
-
-int sys_anon_allocate(size_t size, void **pointer) STUB_IMPL
-
-int sys_anon_free(void *pointer, size_t size) STUB_IMPL
-
-int sys_open(const char *pathname, int flags, int *fd) STUB_IMPL
-
-int sys_read(int fd, void *buf, size_t count, ssize_t *bytes_read) STUB_IMPL
-
-int sys_write(int fd, const void *buf, size_t count, ssize_t *bytes_written) STUB_IMPL
-
-int sys_seek(int fd, off_t offset, int whence, off_t *new_offset) STUB_IMPL
-
-int sys_close(int fd) STUB_IMPL
-
-int sys_vm_map(void *hint, size_t size, int prot, int flags, int fd, off_t offset, void **window) STUB_IMPL
-
-int sys_vm_unmap(void *pointer, size_t size) STUB_IMPL
-
-int sys_clock_get(int clock, time_t *secs, long *nanos) STUB_IMPL
+int sys_tcb_set(void *pointer) STUB_IMPL(sys_tcb_set)
+int sys_futex_wait(int *pointer, int expected, const struct timespec *time) STUB_IMPL(sys_futex_wait)
+int sys_futex_wake(int *pointer) STUB_IMPL(sys_futex_wake)
+int sys_anon_allocate(size_t size, void **pointer) STUB_IMPL(sys_anon_allocate)
+int sys_anon_free(void *pointer, size_t size) STUB_IMPL(sys_anon_free)
+int sys_open(const char *pathname, int flags, int *fd) STUB_IMPL(sys_open)
+int sys_read(int fd, void *buf, size_t count, ssize_t *bytes_read) STUB_IMPL(sys_read)
+int sys_write(int fd, const void *buf, size_t count, ssize_t *bytes_written) STUB_IMPL(sys_write)
+int sys_seek(int fd, off_t offset, int whence, off_t *new_offset) STUB_IMPL(sys_seek)
+int sys_close(int fd) STUB_IMPL(sys_close)
+int sys_vm_map(void *hint, size_t size, int prot, int flags, int fd, off_t offset, void **window) STUB_IMPL(sys_vm_map)
+int sys_vm_unmap(void *pointer, size_t size) STUB_IMPL(sys_vm_unmap)
+int sys_clock_get(int clock, time_t *secs, long *nanos) STUB_IMPL(sys_clock_get)
 
 }
