diff --git a/shr.shr b/shr.shr
index 8f99fc8..ea55faf 100644
--- a/shr.shr
+++ b/shr.shr
@@ -15,7 +15,7 @@ D output_ptr Q output_buf
 D cond_ret R
 
 D clean_exit = 7 0
-D exit = 0 x3C !
+D exit = 0 x0 !
 
 D elf_hdr
     B x7F B 'E  B 'L  B 'F
@@ -151,15 +151,15 @@ D writeerr8
     = 7 2
     = 2 1
     S 0 6
-    = 0 1
+    = 0 x103
     !
     R
 
 D seek_to_start
     = 6 0
 D seek
-    = 2 0
-    = 0 8
+    = 2 3
+    = 0 x104
     = 7 3
     !
     R
@@ -520,7 +520,7 @@ D out_of_code
     L 2 2
     A 6 output_buf
     w 2 6
-    = 0 1
+    = 0 x103
     = 7 3
     !
 
@@ -550,7 +550,7 @@ D out_of_code
     = 2 xB0
     A 6 elf_hdr
     = 7 3
-    = 0 1
+    = 0 x103
     !
 
     J clean_exit
@@ -654,9 +654,9 @@ B 't
 B 0
 
 e
-    = 0 2
+    = 0 x100
     A 7 filename
-    = 6 x141
+    = 6 x13
     = 2 x1FF
     !
     M 0 3
@@ -665,7 +665,7 @@ e
     = 2 xFFFF
     A 6 source_code_buf
     = 7 0
-    = 0 0
+    = 0 x102
     !
 
     = 6 x1000
