From a06f8d8f4f0959b0c06d940d764f5c5515f58477 Mon Sep 17 00:00:00 2001
From: czapek1337 <czapek1337@gmail.com>
Date: Sat, 21 May 2022 21:35:57 +0200
Subject: [PATCH] Add zigux as a target

---
 lib/config.sub | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/config.sub b/lib/config.sub
index d74fb6d..15ea437 100755
--- a/lib/config.sub
+++ b/lib/config.sub
@@ -1723,7 +1723,7 @@ case $os in
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
-	     | os9* | macos* | osx* | ios* \
+	     | os9* | macos* | osx* | ios* | zigux* \
 	     | mpw* | magic* | mmixware* | mon960* | lnews* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* | twizzler* \
-- 
2.36.1

